cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -543,6 +549,51 @@ public Boolean call() throws Exception {
     assertNoAccess(appId);
   }
 
+  @Test
+  public void testCrossNSDatasetAccessWithAuth() throws Exception {
+    createAuthNamespace();
+    getNamespaceAdmin().create(new NamespaceMeta.Builder()
+                                 .setName(CrossNsDatasetAccessApp.DATASET_OUTPUT_SPACE).build());
+    NamespaceId datasetOutputSpace = new NamespaceId(CrossNsDatasetAccessApp.DATASET_OUTPUT_SPACE);
+    addDatasetInstance(datasetOutputSpace.toId(), "keyValueTable", "store");
+    ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.class);
+
+    // Send stream and run the flowlet as BOB. Should fail
+    StreamManager streamManager = getStreamManager(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.STREAM_NAME);
+    for (int i = 0; i < 10; i++) {
+      streamManager.send(String.valueOf(i).getBytes());
+    }
+    Authorizer authorizer = getAuthorizer();
+    authorizer.grant(AUTH_NAMESPACE, BOB, ImmutableSet.of(Action.ALL));
```

Comment:
Since we are testing dataset access. What we expect is that if Bob don't have privileges on the dataset in DATASET_SPACE, it should fail. As a result, we first grant Bob with the privileges of the original namespace with the program to make sure this won't cause the failure.